File: dart/test/rule_local_cursor_normalization_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:test/test.dart';

final Map<String, Object?> _contract =
    jsonDecode(
          File(
            '../capability_conformance/rule_local_cursor_contract.json',
          ).readAsStringSync(),
        )
        as Map<String, Object?>;

void main() {
  test('classifies all contract families by exact authored identity', () {
    for (final row in _rows('family_cases')) {
      final id = row['id']! as String;
      final source = '${row['header']}\n /x/\n';
      final parsed = parseSpec(source);
      final expectedAnd = row['family'] == 'and';

      expect(parsed.rules.single.header.mode.isAnd, expectedAnd, reason: id);
      final compiled = compileSpec(parsed, validateSource: false).rule('Top')!;
      expect(compiled.modeMetadata.isAnd, expectedAnd, reason: id);
    }

    expect(RuleMode.pipe.isAnd, isFalse);
    expect(RuleMode.single.isAnd, isTrue);
  });

  test('parses validates and lowers every contract edge exactly', () {
    final diagnosticRows = {
      for (final row in _rows('diagnostics')) row['code']! as String: row,
    };

    for (final row in _rows('edge_resolution_cases')) {
      final id = row['id']! as String;
      final declared = _strings(row['declared_rules']);
      final source = _edgeSource(row['parent_family']! as String, [
        row['source']! as String,
      ], declared);
      final expectedError = row['expected_error'];

      if (expectedError is String) {
        final diagnostic = _diagnosticFor(source);
        final contractDiagnostic = diagnosticRows[expectedError]!;
        expect(diagnostic.code, expectedError, reason: id);
        expect(diagnostic.stage, contractDiagnostic['stage'], reason: id);
        expect(
          diagnostic.fields.keys.toSet(),
          _strings(contractDiagnostic['fields']).toSet(),
          reason: '$id diagnostic fields',
        );
        expect(diagnostic.field('rule_label'), 'Top', reason: id);
        if (diagnostic.fields.containsKey('target')) {
          expect(
            diagnostic.field('target'),
            expectedError == 'bare_edge_target_undefined' ? 'Missing' : 'Child',
            reason: id,
          );
        }
        if (diagnostic.fields.containsKey('regex_index')) {
          expect(diagnostic.field('regex_index'), 0, reason: id);
        }
        if (diagnostic.fields.containsKey('targets')) {
          expect(diagnostic.field('targets'), declared, reason: id);
        }

        final roundtrip = SpecPortableDiagnostic.fromJson(
          _jsonObject(jsonDecode(jsonEncode(diagnostic.toJson()))),
        );
        expect(roundtrip.toJson(), diagnostic.toJson(), reason: '$id JSON');
        continue;
      }

      final parsed = parseSpec(source);
      validateSpec(parsed);
      final expected = _jsonObject(row['expected']);
      if (expected['kind'] == 'lifecycle') {
        expect(
          parsed.rules.first.body.first.kind,
          isA<LifecycleMarkerBodyElementKind>(),
          reason: id,
        );
        continue;
      }

      if (expected['source_form'] == 'bare') {
        final kind = parsed.rules.first.body.first.kind;
        expect(kind, isA<BareEdgeBodyElementKind>(), reason: id);
        final bare = kind as BareEdgeBodyElementKind;
        final expectedTargets = _objects(expected['targets']);
        expect(bare.targets, hasLength(expectedTargets.length), reason: id);
        expect(bare.code != null, expected['has_block'], reason: id);
        final expectedFluent = expectedTargets.first['fluent'];
        expect(
          bare.fluentChain.firstOrNull == null
              ? null
              : _fluentText(bare.fluentChain.first),
          expectedFluent,
          reason: id,
        );
        expect(
          parsed.rules.first.body
              .map((element) => element.kind)
              .whereType<RawBodyElementKind>(),
          isEmpty,
          reason: '$id must not fall back to raw syntax',
        );

        final roundtrip = SpecFile.fromJson(
          _jsonObject(jsonDecode(jsonEncode(parsed.toJson()))),
        );
        expect(roundtrip.toJson(), parsed.toJson(), reason: '$id AST JSON');
      }

      final compiled = compileSpec(parsed).rule('Top')!;
      final expectedTargets = _objects(expected['targets']);
      if (expected['ownership'] == 'action') {
        expect(
          compiled.actionEdges,
          hasLength(expectedTargets.length),
          reason: id,
        );
        expect(compiled.blindEdges, isEmpty, reason: id);
        for (var index = 0; index < expectedTargets.length; index += 1) {
          final actual = compiled.actionEdges[index].targets.single;
          final target = expectedTargets[index];
          expect(actual.label, target['label'], reason: id);
          expect(actual.index, target['index'] ?? 0, reason: id);
        }
      } else {
        expect(
          compiled.blindEdges,
          hasLength(expectedTargets.length),
          reason: id,
        );
        expect(compiled.actionEdges, isEmpty, reason: id);
        for (var index = 0; index < expectedTargets.length; index += 1) {
          expect(
            compiled.blindEdges[index].target.label,
            expectedTargets[index]['label'],
            reason: id,
          );
        }
      }
    }
  });

  test('validates normalized ownership for all contract edge sets', () {
    for (final row in _rows('rule_edge_set_cases')) {
      final id = row['id']! as String;
      final source = _edgeSource(
        row['parent_family']! as String,
        _strings(row['sources']),
        _strings(row['declared_rules']),
      );
      final expectedError = row['expected_error'];
      if (expectedError is String) {
        final diagnostic = _diagnosticFor(source);
        expect(diagnostic.code, expectedError, reason: id);
        expect(diagnostic.stage, 'validate_rule', reason: id);
        expect(diagnostic.fields.keys.toSet(), {
          'ownerships',
          'rule_label',
        }, reason: id);
        expect(diagnostic.field('ownerships'), ['action', 'blind'], reason: id);
        continue;
      }

      final compiled = compileSpec(parseSpec(source)).rule('Top')!;
      if (row['expected_ownership'] == 'action') {
        expect(compiled.actionEdges, isNotEmpty, reason: id);
        expect(compiled.blindEdges, isEmpty, reason: id);
      } else {
        expect(compiled.blindEdges, isNotEmpty, reason: id);
        expect(compiled.actionEdges, isEmpty, reason: id);
      }
    }
  });

  test('retains only complete-line and header-rest bare candidates', () {
    for (final source in [
      'Top::\n Child\n\nChild:\n /x/\n',
      'Top:: Child\n\nChild:\n /x/\n',
      'Top::AND\n Child {\n  return(child_result)\n }\n\nChild:\n /x/\n',
    ]) {
      final parsed = parseSpec(source);
      expect(
        parsed.rules.first.body.first.kind,
        isA<BareEdgeBodyElementKind>(),
      );
      expect(
        parsed.rules.first.body
            .map((element) => element.kind)
            .whereType<RawBodyElementKind>(),
        isEmpty,
      );
    }

    final suffix = parseSpec('Top:: /x/ Child\n\nChild:\n /x/\n');
    expect(
      suffix.rules.first.body
          .map((element) => element.kind)
          .whereType<BareEdgeBodyElementKind>(),
      isEmpty,
      reason: 'bare recognition is physical-line scoped',
    );
  });

  test(
    'keeps runtime-family and generated-v1 boundaries explicitly staged',
    () {
      final pipe = compileSpec(parseSpec('Top::|\n /x/\n')).rule('Top')!;
      expect(pipe.modeMetadata.isAnd, isFalse);
      expect(pipe.modeMetadata.usesLegacyAndInterpretation, isTrue);
      expect(
        classifyGeneratedRuleFamily(pipe),
        GeneratedRuleFamily.andSingleAcode,
        reason: 'generated-source v1 migration is owned by .9.1.5.4',
      );
    },
  );
}

List<Map<String, Object?>> _rows(String name) {
  return _objects(_contract[name]);
}

List<Map<String, Object?>> _objects(Object? value) {
  return (value! as List)
      .map((item) => Map<String, Object?>.from(item! as Map))
      .toList(growable: false);
}

List<String> _strings(Object? value) {
  return (value! as List).cast<String>();
}

Map<String, Object?> _jsonObject(Object? value) {
  return Map<String, Object?>.from(value! as Map);
}

String _edgeSource(
  String parentFamily,
  List<String> sources,
  List<String> declaredRules,
) {
  final buffer = StringBuffer(parentFamily == 'and' ? 'Top::AND\n' : 'Top::\n');
  for (final source in sources) {
    buffer.writeln(' $source');
  }
  for (final label in declaredRules) {
    buffer.write('\n$label:\n /x/ /y/\n');
  }
  return buffer.toString();
}

SpecPortableDiagnostic _diagnosticFor(String source) {
  try {
    validateSpec(parseSpec(source));
  } on SpecValidationException catch (error) {
    final diagnostic = error.diagnostic;
    if (diagnostic == null) {
      fail('expected portable diagnostic, got: ${error.message}');
    }
    return diagnostic;
  }
  fail('expected spec validation to fail');
}

String _fluentText(FluentCall call) {
  return call.args.isEmpty ? call.method : '${call.method}(${call.args})';
}

File: julia/test/rule_local_cursor_normalization_test.jl
const RULE_LOCAL_CURSOR_CONTRACT = JSON3.read(
    read(
        joinpath(REPO_ROOT, "capability_conformance", "rule_local_cursor_contract.json"),
        String,
    ),
    Dict{String,Any},
)

function _cursor_contract_rows(name::AbstractString)
    return RULE_LOCAL_CURSOR_CONTRACT[String(name)]
end

function _cursor_edge_source(parent_family, sources, declared_rules)
    io = IOBuffer()
    println(io, parent_family == "and" ? "Top::AND" : "Top::")
    for source in sources
        println(io, " ", source)
    end
    for label in declared_rules
        println(io)
        println(io, label, ":")
        println(io, " /x/ /y/")
    end
    return String(take!(io))
end

function _cursor_diagnostic(source::AbstractString)
    try
        validate_spec(parse_spec(source))
    catch error
        if error isa SpecValidationException && error.diagnostic !== nothing
            return error.diagnostic
        end
        rethrow()
    end
    error("expected portable spec validation diagnostic")
end

function _cursor_fluent_text(call::FluentCall)
    return isempty(call.args) ? call.method : "$(call.method)($(call.args))"
end

@testset "Neutral rule-local cursor normalization contract" begin
    family_rows = _cursor_contract_rows("family_cases")
    edge_rows = _cursor_contract_rows("edge_resolution_cases")
    edge_set_rows = _cursor_contract_rows("rule_edge_set_cases")
    diagnostic_rows = Dict(row["code"] => row for row in _cursor_contract_rows("diagnostics"))

    @test length(family_rows) == 36
    @test length(edge_rows) == 18
    @test length(edge_set_rows) == 6

    @testset "all authored rule families" begin
        for row in family_rows
            parsed = parse_spec("$(row["header"])\n /x/\n")
            mode = only(parsed.rules).header.mode
            expected_family = row["family"]
            expected_policy = row["cursor_policy"]

            @test rule_family(mode) == expected_family
            @test cursor_policy(mode) == expected_policy

            compiled = compiled_rule(compile_spec(parsed; validate_source = false), "Top")
            @test compiled !== nothing
            @test rule_family(compiled.mode_metadata) == expected_family
            @test cursor_policy(compiled.mode_metadata) == expected_policy
        end

        @test !is_and(RuleMode("Pipe"))
        @test is_and(RuleMode("Single"))
    end

    @testset "every edge form and diagnostic" begin
        for row in edge_rows
            id = row["id"]
            declared = String[String(value) for value in row["declared_rules"]]
            source = _cursor_edge_source(
                row["parent_family"],
                [String(row["source"])],
                declared,
            )

            expected_error = get(row, "expected_error", nothing)
            if expected_error !== nothing
                diagnostic = _cursor_diagnostic(source)
                contract_diagnostic = diagnostic_rows[expected_error]
                @test diagnostic.code == expected_error
                @test diagnostic.stage == contract_diagnostic["stage"]
                @test Set(keys(diagnostic.fields)) == Set(String(value) for value in contract_diagnostic["fields"])
                @test diagnostic.fields["rule_label"] == "Top"
                if haskey(diagnostic.fields, "target")
                    @test diagnostic.fields["target"] == (
                        expected_error == "bare_edge_target_undefined" ? "Missing" : "Child"
                    )
                end
                if haskey(diagnostic.fields, "regex_index")
                    @test diagnostic.fields["regex_index"] == 0
                end
                if haskey(diagnostic.fields, "targets")
                    @test diagnostic.fields["targets"] == declared
                end

                roundtrip = from_json(
                    SpecPortableDiagnostic,
                    JSON3.read(JSON3.write(to_json(diagnostic))),
                )
                @test to_json(roundtrip) == to_json(diagnostic)
                continue
            end

            parsed = parse_spec(source)
            validate_spec(parsed)
            expected = row["expected"]
            first_kind = first(first(parsed.rules).body).kind

            if expected["kind"] == "lifecycle"
                @test first_kind isa LifecycleMarkerBodyElementKind
                @test first_kind.marker == expected["name"]
                continue
            end

            if expected["source_form"] == "bare"
                @test first_kind isa BareEdgeBodyElementKind
                @test length(first_kind.targets) == length(expected["targets"])
                @test (first_kind.code !== nothing) == expected["has_block"]
                expected_fluent = first(expected["targets"])["fluent"]
                actual_fluent = isempty(first_kind.fluent_chain) ? nothing : _cursor_fluent_text(first(first_kind.fluent_chain))
                @test actual_fluent == expected_fluent
                @test all(element -> !(element.kind isa RawBodyElementKind), first(parsed.rules).body)

                roundtrip = from_json(SpecFile, JSON3.read(JSON3.write(to_json(parsed))))
                @test to_json(roundtrip) == to_json(parsed)
            end

            compiled = compiled_rule(compile_spec(parsed), "Top")
            @test compiled !== nothing
            expected_targets = expected["targets"]
            if expected["ownership"] == "action"
                @test length(compiled.action_edges) == length(expected_targets)
                @test isempty(compiled.blind_edges)
                for (index, target) in enumerate(expected_targets)
                    actual = only(compiled.action_edges[index].targets)
                    @test actual.label == target["label"]
                    @test actual.index == something(target["index"], 0)
                end
            else
                @test length(compiled.blind_edges) == length(expected_targets)
                @test isempty(compiled.action_edges)
                for (index, target) in enumerate(expected_targets)
                    @test compiled.blind_edges[index].target.label == target["label"]
                end
            end
        end
    end

    @testset "normalized ownership sets" begin
        for row in edge_set_rows
            declared = String[String(value) for value in row["declared_rules"]]
            sources = String[String(value) for value in row["sources"]]
            source = _cursor_edge_source(row["parent_family"], sources, declared)
            expected_error = get(row, "expected_error", nothing)

            if expected_error !== nothing
                diagnostic = _cursor_diagnostic(source)
                @test diagnostic.code == expected_error
                @test diagnostic.stage == "validate_rule"
                @test Set(keys(diagnostic.fields)) == Set(["ownerships", "rule_label"])
                @test diagnostic.fields["ownerships"] == ["action", "blind"]
                continue
            end

            compiled = compiled_rule(compile_spec(parse_spec(source)), "Top")
            @test compiled !== nothing
            if row["expected_ownership"] == "action"
                @test !isempty(compiled.action_edges)
                @test isempty(compiled.blind_edges)
            else
                @test !isempty(compiled.blind_edges)
                @test isempty(compiled.action_edges)
            end
        end
    end

    @testset "bare candidates are physical-line scoped" begin
        for source in [
            "Top::\n Child\n\nChild:\n /x/\n",
            "Top:: Child\n\nChild:\n /x/\n",
            "Top::AND\n Child {\n  return(child_result)\n }\n\nChild:\n /x/\n",
        ]
            parsed = parse_spec(source)
            @test first(first(parsed.rules).body).kind isa BareEdgeBodyElementKind
            @test all(element -> !(element.kind isa RawBodyElementKind), first(parsed.rules).body)
        end

        suffix = parse_spec("Top:: /x/ Child\n\nChild:\n /x/\n")
        @test all(element -> !(element.kind isa BareEdgeBodyElementKind), first(suffix.rules).body)
    end
end
